[#65994] PDF export of project list

as-op opened the pull request:
# Ticket

https://community.openproject.org/work_packages/65994

# What are you trying to achieve?

MVP: Implement PDF export for the project list

# What approach did you choose and why?

Extracted the generic part out of the BMDS branch, all hardcoded features are extracted into their own feature work packages.

- [x] PDF Icon in export modal
- [x] PDF generation in a worker job
- [x] Stay on the project list page and display a job progress dialog (with turbo)
- [x] Table of content PDF page shows the hierarchy of the exported projects 
        - if there is more than one project in the query, no TOC page exported otherwise
- [x] Index numbers by hierarchy
- [x] Default cover page 
- [x] Project titles are linked to the respective project URL
- [x] Attributes are displayed
       - in the order and selection of the project list table
       - if not empty
       - Column "Favorite" is not exported
       - Column "Name" is not exported as attribute but as a section title

# Merge checklist

- [x] Added/updated tests
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

as-op committed:
[#65994] PDF export of project list

https://community.openproject.org/work_packages/65994

as-op committed:
Merge branch 'dev' into feature/65994-pdf-export-of-project-list

as-op committed:
add tests

as-op committed:
obey rubocop

as-op committed:
Merge branch 'dev' into feature/65994-pdf-export-of-project-list

as-op committed:
remove not (yet) used work package table code

as-op committed:
adjust for dynamic value in test

EinLama commented on app/models/projects/exports/formatters/active.rb:
```diff
@@ -1,5 +1,3 @@
-# frozen_string_literal: true
```
Why did you remove this? Did it cause issues? If not, please leave it in.

as-op replied:
Probably be mistake cherry picking from the BMDS branch. I wonder why rubocop did not complain as we have a rule for it. Thanks! I place it back.

EinLama commented on app/models/projects/exports/formatters/favored.rb:
```diff
@@ -0,0 +1,44 @@
+      ##
+      # Takes a project and returns yes/no depending on the favored attribute
+      def format(project, **)
+        project.favored_by?(User.current) ? I18n.t(:general_text_Yes) : I18n.t(:general_text_No)
```
I know that this uppercase variant of the translation key is used in other places, too. What I don't know is: why do we do that? It is inconsistent as we go with snake_case for all other i18n keys. We even have both keys in our YML file:

* `general_text_Yes`
* `general_text_yes`

Both are used in different places of the application. 😅 It is not specifically the responsibility of this PR to fix it _everywhere_. But it would be nice to use only the lowercase variant in as many places as possible and this would be a good place to start.

Same applies to keys ending with "No", too.

as-op replied:
> You are right, the uppercase versions are ugly. :D I change it!

Follow up: The actual values of those keys are different. It's also capitalized `Yes` vs. lowercase `yes`. As capitalizing can be tricky in some languages, e.g. additional letters and such, I rather leave that task for the translators than use `capitalize `, `upcase_first` or  `titleize`.

EinLama commented on app/models/projects/exports/formatters/active.rb:
```diff
@@ -31,7 +29,7 @@ module Projects::Exports
     class Active < ::Exports::Formatters::Default
       def self.apply?(attribute, export_format)
-        export_format == :pdf && %i[active].include?(attribute.to_sym)
+        export_format == :pdf && attribute.to_sym == :active
```
The base class implements the following method:

```ruby
      def self.key
        name.demodulize.underscore.to_sym
      end
```

This means, instead of having to write a very similar method for every formatter of this PR, we could add the following implementation to the base class:

```ruby
      def self.apply?(attribute, export_format)
        export_format == :pdf && attribute.to_sym == self.key
      end
```

With that, we don't have to spell it out for every file 😄

as-op replied:
Interesting, but I would rather not change it. This function is called a lot, I keep the constant, so the effort is less. 

Also, rubocop is not happy with the usage of`self` and `attribute.to_sym == key` looks like I forgot a parameter and not so much as a class function. 😄

EinLama commented on app/models/projects/exports/pdf.rb:
```diff
@@ -0,0 +1,141 @@
+    include Exports::PDF::Common::Attachments
+    include Exports::PDF::Common::Markdown
+    include Exports::PDF::Components::Page
+    include Exports::PDF::Components::WpTable
```
Is this include needed? It does not seem like any of the included methods is being used.

Is this already a preparation for the follow up for work packages?

as-op replied:
Good catch, no it's not yet necessary. I will reappear for the follow-up work, but for now I remove it. Thanks!

EinLama commented on app/models/projects/exports/pdf.rb:
```diff
@@ -0,0 +1,141 @@
+    end
+
+    def title
+      build_pdf_filename([heading].join("_"))
```
`heading` is always a single string. Why the added Array and `join`? To catch `nil` values and render `_` instead? This would be an alternative for that use case:

```ruby
build_pdf_filename(heading || "_")
```
 
Such a guard could also be moved into the `build_pdf_filename` method, depending on how it is intended to be called.

as-op replied:
LOL, good catch! There were more elements in that array at first, but this is pointless ^^
I change it to `build_pdf_filename(heading)`

EinLama commented on app/models/projects/exports/pdf_export/info_map.rb:
```diff
@@ -0,0 +1,80 @@
+      infos_map = {}
+      projects.each_with_index do |project, index|
+        infos_map[project.id] = { level_path: [index + 1], level: 0, children: [], project: }
+      end
```
This portion could be simplified to:

```suggestion
      infos_map = projects.map.with_index do |project, index|
        [project.id, level_path: [index + 1], level: 0, children: [], project:]
      end.to_h
```

as-op replied:
Much nicer! 👍

EinLama commented on app/models/projects/exports/pdf_export/info_map.rb:
```diff
@@ -0,0 +1,80 @@
+    end
+
+    def init_meta_infos_map_nodes(projects)
+      infos_map = {}
+      projects.each do |project|
+        infos_map[project.id] = { level_path: [], level: 0, children: [], project: }
+      end
+      infos_map
```
```suggestion
      projects.map do |project|
        [project.id, level_path: [], level: 0, children: [], project:]
      end.to_h
```

as-op replied:
rubocop directed me to an even more compact version
```
projects.to_h do |project|
        [project.id, { level_path: [], level: 0, children: [], project: }]
end
```

EinLama replied:
I didn't know about that, nice! :)

EinLama commented on app/models/projects/exports/pdf_export/report.rb:
```diff
@@ -0,0 +1,255 @@
+    end
+
+    def selects
+      @selects = query
```
The instance variable does not seem to be used anywhere. It could be removed -> less state. However, this method is also called multiple times in rather quick succession. Consider caching it:

```ruby
@selects ||= query
          # ...
```

as-op replied:
`||=` was what I was going for. Thanks! 
But I remove the caching and just use a local variable in the function where it is called twice.

EinLama commented on app/models/projects/exports/pdf_export/report.rb:
```diff
@@ -0,0 +1,255 @@
+      return false if attribute.nil? || %i[name favored].include?(attribute)
+
+      true
+    end
```
I think the guard clause harms more than it helps in such a small method. Consider this:

```suggestion
    def can_view_attribute?(_project, attribute)
      attribute && !%i[name favored].include?(attribute)
    end
```

as-op replied:
Agreed, I remove the pointless guard.

EinLama commented on frontend/src/stimulus/controllers/dynamic/job-dialog.controller.ts:
```diff
@@ -0,0 +1,112 @@
+      }
+    }
+
+    async requestJob():Promise<string> {
```
The older exports have this issue, too. But since this is a new dialog controller, here we go:

If I refrain from starting GoodJob, the PDF export job will never get picked up. Since we define no timeout here, I will forever see the export modal with the animated progress bar - even after minutes of hopelessness.

Did you consider adding a sane timeout here so that users don't wait for forever in case that GoodJob has crashed?

as-op replied:
Good point, but I can't do something about it here. The function does not wait as it gets an ID whether GoodJob runs or not. It then calls `/job_statuses/${job_id}/dialog` and in this we would need to time out (or rather in the backend where the turbo frame is updated). But that is out of scope for this PR.

EinLama replied:
Ah, I see! Thanks for the explanation. I agree that this is out of scope for this PR.

EinLama commented on frontend/src/stimulus/controllers/dynamic/job-dialog.controller.ts:
```diff
@@ -0,0 +1,112 @@
+                if (job_id) {
+                    return this.showJobModal(job_id);
+                }
+                return this.handleError('No job ID returned from server.');
```
I think these error messages should be translated. Hopefully, no user will ever see them. But if they do, it should be in their language of choice.

as-op replied:
Right! I found two more untranslated ones and fixed them, too. Thanks!

EinLama commented on frontend/src/stimulus/controllers/dynamic/job-dialog.controller.ts:
```diff
@@ -0,0 +1,112 @@
+      const dialog = document.getElementById(this.closeDialogIdValue) as HTMLDialogElement;
+      if (dialog) {
+        dialog.close();
+      }
```
Could be:

```suggestion
      dialog?.close();
```

EinLama reviewed (changes requested):
Well done! I tested multiple edge cases and failed to break the export 👏

I still do have a few questions, please have a look and let me know what you think.

I have also left some remarks and code suggestions. Most of these are optional and subject to personal preference, feel free to pick any of them that you like.

as-op committed:
close code review issues & hints
https://github.com/opf/openproject/pull/19756#discussion_r2273316765
https://github.com/opf/openproject/pull/19756#discussion_r2273341315
https://github.com/opf/openproject/pull/19756#discussion_r2273363760
https://github.com/opf/openproject/pull/19756#discussion_r2273377147
https://github.com/opf/openproject/pull/19756#discussion_r2273402627
https://github.com/opf/openproject/pull/19756#discussion_r2273405467
https://github.com/opf/openproject/pull/19756#discussion_r2273443567
https://github.com/opf/openproject/pull/19756#discussion_r2273470141

EinLama commented on app/models/projects/exports/pdf_export/report.rb:
```diff
@@ -167,9 +166,7 @@ def table_entry(project, value_name, caption)
-      return false if attribute.nil? || %i[name favored].include?(attribute)
-
-      true
+      attribute.nil? || %i[name favored].include?(attribute)
```
I think this code change inverted the boolean logic from before. It should be:

```ruby
attribute && !%i[name favored].include?(attribute)
```

as-op replied:
Waaaah! You're right! 😳

as-op committed:
close code review issues & hints; obey new eslint rules
https://github.com/opf/openproject/pull/19756#discussion_r2273517650
https://github.com/opf/openproject/pull/19756#discussion_r2273520150

as-op committed:
close code review issues & hints
https://github.com/opf/openproject/pull/19756#discussion_r2275892911

EinLama reviewed (approved):
Nice 🎉 Thanks for adjusting some of the stuff.

Feel free to merge as soon as the CI is green. :shipit:

as-op committed:
obey eslint

as-op committed:
Merge branch 'dev' into feature/65994-pdf-export-of-project-list
